src/FFmpegWriter.cpp

```diff
@@ -1361,23 +1352,51 @@ void FFmpegWriter::open_video(AVFormatContext *oc, AVStream *st) {
 
 #if IS_FFMPEG_3_2
 	if (hw_en_on && hw_en_supported) {
-		video_codec->max_b_frames = 0;        // At least this GPU doesn't support b-frames
-		video_codec->pix_fmt   = hw_en_av_pix_fmt;
-		video_codec->profile = FF_PROFILE_H264_BASELINE | FF_PROFILE_H264_CONSTRAINED;
-		av_opt_set(video_codec->priv_data,"preset","slow",0);
-		av_opt_set(video_codec->priv_data,"tune","zerolatency",0);
-		av_opt_set(video_codec->priv_data, "vprofile", "baseline", AV_OPT_SEARCH_CHILDREN);
+		switch (video_codec->codec_id) {
+			case AV_CODEC_ID_H264:
+				video_codec->max_b_frames = 0;        // At least this GPU doesn't support b-frames
+				video_codec->pix_fmt   = hw_en_av_pix_fmt;
+				video_codec->profile = FF_PROFILE_H264_BASELINE | FF_PROFILE_H264_CONSTRAINED;
+				av_opt_set(video_codec->priv_data, "preset", "slow", 0);
+				av_opt_set(video_codec->priv_data, "tune", "zerolatency", 0);
+				av_opt_set(video_codec->priv_data, "vprofile", "baseline", AV_OPT_SEARCH_CHILDREN);
+				break;
+			case AV_CODEC_ID_HEVC:
+				// must be zero (at least on AMDGPU on Linux 8/2019)
+				video_codec->bit_rate = 0;
```

Comment:
Hrrrm. This is the only part I'm _really_ unsure about. (Along with the preset in OpenShot/openshot-qt#2941 that defines High, Medium, and Low quality settings with identical 0 bitrates — though that's admittedly somewhat our fault for having such an inflexible preset-definition syntax.) It's odd that bitrate wouldn't work, too, since the documentation [_seems_ to indicate](http://www.ffmpeg.org/ffmpeg-codecs.html#VAAPI-encoders) that it's supported... I guess maybe this is part of VAAPI on AMD [only having "partial" support](https://trac.ffmpeg.org/wiki/HWAccelIntro#PlatformAPIAvailability), perhaps?

Still... I don't know if you saw that OpenShot _also_ has CRF support, as an alternative to raw bitrates — we're [already using it](https://github.com/OpenShot/openshot-qt/blob/develop/src/presets/format_mp4_x265_crf.xml) for software-encoded HEVC in the x265 preset, in fact.

Unfortunately, fat lot of good _that_ does us, either, because [VAAPI doesn't support that, either](https://trac.ffmpeg.org/wiki/Hardware/VAAPI#Mappingoptionsfromlibx264).
> No CRF-like mode is currently supported. The only constant-quality mode is CQP (constant quantisation parameter), which has no adaptivity to scene content. It does, however, allow different quality settings for different frame types, to improve compression by spending fewer bits on unreferenced B-frames - see the `(i|b)_q(factor|offset)` options.

However, here again it _seems_ to imply that setting a bitrate is supported and will use VBR by default — though it does say that it can't be combined with a `qp` setting... Just out of curiosity, when you tested setting `bit_rate`, were you _**also**_ setting `qp`? Sounds like that may be why it wouldn't work (emphasis mine):
> **CQP mode cannot be combined with a maximum bitrate or buffer size.**
>
> CBR and VBR modes are supported, though the output of them varies significantly by driver and device (default is VBR, set `-maxrate` equal to `-b:v` for CBR). HRD buffering options (rc_max_rate, rc_buffer_size) are functional, and the encoder will generate buffering_period and pic_timing SEI when appropriate.

They also give some examples:
> Encode an input with H.264 at 5Mbps VBR:
> ```
> ffmpeg -vaapi_device /dev/dri/renderD128 -i input.mp4 -vf 'format=nv12,hwupload' -c:v h264_vaapi -b:v 5M output.mp4
> ```

> Encode with H.264 at good constant quality:
> ```
> ffmpeg -vaapi_device /dev/dri/renderD128 -i input.mp4 -vf 'format=nv12,hwupload' -c:v h264_vaapi -qp 18 output.mp4
> ```
> Encode with 10-bit H.265 at 15Mbps VBR (recent hardware required - Kaby Lake or later Intel):
> ```
> ffmpeg -vaapi_device /dev/dri/renderD128 -i input.mp4 -vf 'format=p010,hwupload' -c:v hevc_vaapi -b:v 15M -profile 2 output.mp4
> ```

If setting bitrate can't be made to work, I think we need to provide _some_ way of configuring the quality targets so we have real High, Medium, and Low options. Maybe it's time to implement a `cqp` bitrate mode that sets `qp`, to go along with the `crf` support here:
https://github.com/OpenShot/libopenshot/blob/857297abdf3b3c4d8067b5f32559d10269b2d7e0/src/FFmpegWriter.cpp#L399-L463

Reply:
This is a bit problematic! I started with those example ffmpeg commands (with -b:v 5M) but it produced garbage. I tried that on two Linux machines (gentoo-based and arch-based) with open-source AMDGPU driver and two different AMD cards. Setting bitrate to 0 and modifying qp did the trick. Do you think bitrate has to work? Maybe there is a driver or encoder bug?

Original: https://k3a.me/u/big_buck_bunny_720p_1mb.mp4
Encoded: https://k3a.me/u/bunny_bug.mp4

Encoding done with
`fmpeg -vaapi_device /dev/dri/renderD128 -i big_buck_bunny_720p_1mb.mp4 -vf 'format=nv12,hwupload' -c:v hevc_vaapi -qp:v 25 -b:v 5M -y bunny_bug.mp4`

Not setting `-qp:v 25` (but keeping -b:v 5M) resulted in another garbage. :( 
Here: https://k3a.me/u/bunny_bug2.mp4

GPU Radeon RX 550 (POLARIS11), ffmpeg n4.1.3, libva-mesa-driver-19.1.2-1

From the [ffmpeg source](https://github.com/FFmpeg/FFmpeg/blob/20fed2f0ab197d60801280dfc844f6b29a397ff2/libavcodec/vaapi_encode.c#L1479) it looks like there are several modes. I will try to look into it and make it select a different mode on Monday/Tuesday and see if there is a way to use bit_rate on my HW and report back.

Reply:
> This is a bit problematic!

Yeah, it's really a pain. I'm hoping we can lick it with enough trial-and-error, though it's sad that it always seems to come down to that with ffmpeg. I'll do some stackoverflow-and-etc. scouring over the weekend as well, though.

One other test, if you'd be willing? The same no-QP encode, but also setting `-maxrate` as they'd suggested, for _fixed_ bitrate. I'm curious whether _that_ might at least work:
```
ffmpeg -vaapi_device /dev/dri/renderD128 -i big_buck_bunny_720p_1mb.mp4 \
-vf 'format=nv12,hwupload' -c:v hevc_vaapi -b:v 5M -maxrate 5M -y bunny_bug.mp4
```

Reply:
Unfortunately the same - https://k3a.me/u/bunny_bug3.mp4 :(

Reply:
Poop. Oh well. But setting `-qp:v` works fine? Weird. Do you need to _explicitly_ set bitrate to 0, or can you just omit it? 

If an explicit bitrate setting can be omitted entirely, then I'm leaning back towards my "cqp mode" fix. That wouldn't be so hard to implement, it's basically a cut-and-paste mod of the existing crf code. So, that's an option, I'll think about putting it together (would be good to have anyway) and bounce the idea off @jonoomph . (...Consider it bounced! **#tag**)

Reply:
Sorry, I had to focus on something else past week but I am ready to work on this tomorrow.

I did some ffmpeg cmdline testing today after updating the system and it seems that bitrate setting now magically works in the updated `libva-mesa-driver-19.1.5-1`. It works for AMD but doesn't for Intel (well it at least produces a working file but with 76 kb/s bitrate).

Reply:
Hm. You lost me there, which hardware produces a file with 76kbps bitrate, and under what conditions?

Reply:
Sorry, I will try to explain.

In the meantime, I updated my Arch Linux box and there were new versions of FFMpeg and Mesa. I also rebooted and enabled iGPU (internal GPU in the CPU) as I realized it should support VP9 too. So now I have two VAAPI HW encoders. And with those updated libraries, HW encoding started behaving better - it made usable files using AMD as well as Intel but Intel had 76 kb/s bitrate (despite using the same ffmpeg comamndline options).

It is much easier to look at my test script here - https://gist.github.com/k3a/dcfb3c34745899c50eb1bb0967c0f3be those commented-out lines were also tested. Comments preceding a line belongs to that line.

I've spent almost the whole day today on this and found this out:
1. it is possible to ask ffmpeg which options an encoder support (`ffmpeg -h encoder=h264_vaapi`). The problem is, those seem to be codec-specific but encoding also accepts some (which?!) of the common ones, like bitrate, of course..
2. so `vaapi` encoders all support `rc_mode` which is so called explicit resource control mode
3. ffmpeg's `libavcodec/vaapi_encode.c` decides which rc_mode to apply depending on various combinations and it looks like it can be wrong sometimes and those rules can change with each ffmpeg version bump
4. if I set rc_mode explicitly, it started behaving consistently, except for "VBR" mode, where `maxrate` had to be set as well.
5. I had to enable Ffmpeg log while working on it. It really gives very helpful output if some codec configuration doesn't work so it will be really good to have later (not part of this commit). For example:
```
ffmpeg: Input surface format is nv12.
ffmpeg: Compatible profile VAProfileVP9Profile0 (19) is not supported by driver.
ffmpeg: No usable encoding profile found.
```
6. There needs to be proper HW card detection made in openshot-qt. Currently it allows no acceleration or card 1. I have two cards now, though. On linux, it is quite easy to do, just look for dev files `ls -1 /dev/dri/renderD12*`:
```
/dev/dri/renderD128 # that is my AMD
/dev/dri/renderD129 # that is my Intel
```
maybe there is some better way to detect it or even get GPU name, I don't know yet. But I know you can do `vainfo --display drm --device /dev/dri/renderD128 | head -n3` to get:
```
vainfo: VA-API version: 1.5 (libva 2.5.0)
vainfo: Driver version: Mesa Gallium driver 19.1.5 for Radeon RX 560 Series (POLARIS11, DRM 3.32.0, 5.2.9-arch1-1-ARCH, LLVM 8.0.1)
vainfo: Supported profile and entrypoints
```
7. original property setting for h.264 HW encoding may be useless. Unfortunately it is hard to find proper docs for ffmpeg but here is what I think about those:
```c
// probably nvenc only (Fmpeg/libavcodec/nvenc_h264.c)
av_opt_set(video_codec->priv_data, "preset", "slow", 0);
// seems like non-sense. It is top level bool parameter (Fmpeg/libavcodec/nvenc_h264.c)
// should probably be av_opt_set(video_codec->priv_data, "zerolatency", "1", 0);
av_opt_set(video_codec->priv_data, "tune", "zerolatency", 0);
// probably nonsense. Couldn't grep-find vprofile in the whole ffmpeg source; I think only generic video_codec->profile is used here
av_opt_set(video_codec->priv_data, "vprofile", "baseline", AV_OPT_SEARCH_CHILDREN);
```
I kept it as it was there though; in order to prevent breaking previously-working code.

I've updated the code and also included your cqp commit. But I haven't tested cqp yet. 
I tested that H264, HEVC and VP9 produce usable results. 
For 15 Mb/s default OpeNshot export settings it produced:

**AMD tests**:
H.264 9062 kb/s
HEVC 15002 kb/s
VP9 unsupported

**Intel test**:
H.264 9202 kb/s
HEVC 15252 kb/s
VP9 15778 kb/s

Any comments or problems with the pull requests?